File: src/print_partner/core/parsers.py
"""STL filename/path parsers ported from stl-manifest-generator."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum
from pathlib import PurePosixPath


class PartRole(str, Enum):
    PRIMARY = "primary"
    ACCENT = "accent"
    CLEAR = "clear"
    OPAQUE = "opaque"


ROLE_MARKERS = {
    "[a]": PartRole.ACCENT,
    "[c]": PartRole.CLEAR,
    "[o]": PartRole.OPAQUE,
}

QUANTITY_RE = re.compile(r"[ _]x([0-9]+)\.stl$", re.IGNORECASE)
ROLE_PREFIX_RE = re.compile(r"^\[[aco]\]", re.IGNORECASE)


@dataclass(frozen=True)
class ParsedPart:
    role: PartRole
    quantity: int
    part_slug: str
    filename: str


def _check_role_in_text(text: str) -> PartRole | None:
    lower = text.lower()
    for marker, role in ROLE_MARKERS.items():
        if marker in lower:
            return role
    return None


def parse_role(path_or_name: str) -> PartRole:
    """Detect role from path segments and filename ([a]/[c]/[o])."""
    posix = PurePosixPath(path_or_name.replace("\\", "/"))
    for segment in (*posix.parent.parts, posix.name):
        found = _check_role_in_text(segment)
        if found is not None:
            return found
    return PartRole.PRIMARY


def parse_quantity(filename: str) -> int:
    m = QUANTITY_RE.search(filename)
    if m:
        return max(1, int(m.group(1)))
    return 1


def parse_part_slug(filename: str) -> str:
    """Basename stripped of role prefixes and _xN quantity suffix."""
    name = PurePosixPath(filename).name
    if name.lower().endswith(".stl"):
        stem = name[:-4]
    else:
        stem = name
    stem = ROLE_PREFIX_RE.sub("", stem)
    stem = re.sub(r"[ _]x[0-9]+$", "", stem, flags=re.IGNORECASE)
    return stem or name


def parse_stl_path(relative_path: str) -> ParsedPart:
    filename = PurePosixPath(relative_path.replace("\\", "/")).name
    return ParsedPart(
        role=parse_role(relative_path),
        quantity=parse_quantity(filename),
        part_slug=parse_part_slug(filename),
        filename=filename,
    )

File: web/packages/domain/src/parsers.ts
/** STL filename conventions (ported from Python parsers.py). */

import type { NamingProfile } from "./stl-naming.js";
import { getDefaultNamingProfile } from "./stl-naming.js";

export type PartRole = "primary" | "accent" | "clear" | "opaque";

export type ParsedPart = {
  role: PartRole;
  quantity: number;
  partSlug: string;
  filename: string;
};

function resolveProfile(profile?: NamingProfile | null): NamingProfile {
  return profile ?? getDefaultNamingProfile();
}

function roleFromId(roleId: string): PartRole {
  if (roleId === "accent" || roleId === "clear" || roleId === "opaque") return roleId;
  return "primary";
}

function checkRoleInText(text: string, profile: NamingProfile): PartRole | null {
  const lower = text.toLowerCase();
  for (const [marker, roleId] of profile.markerRoleMap) {
    if (lower.includes(marker.toLowerCase())) return roleFromId(roleId);
  }
  return null;
}

function folderRole(relativePath: string, profile: NamingProfile): PartRole | null {
  const posix = relativePath.replace(/\\/g, "/").toLowerCase();
  for (const rule of profile.folderRules) {
    if (posix.includes(rule.path_contains.toLowerCase())) {
      return roleFromId(rule.role_id);
    }
  }
  return null;
}

function basename(path: string): string {
  const normalized = path.replace(/\\/g, "/");
  const idx = normalized.lastIndexOf("/");
  return idx >= 0 ? normalized.slice(idx + 1) : normalized;
}

function parentParts(path: string): string[] {
  const normalized = path.replace(/\\/g, "/");
  const parts = normalized.split("/");
  parts.pop();
  return parts;
}

export function parseRole(pathOrName: string, profile?: NamingProfile | null): PartRole {
  const prof = resolveProfile(profile);
  const foundFolder = folderRole(pathOrName, prof);
  if (foundFolder) return foundFolder;
  const segments = [...parentParts(pathOrName), basename(pathOrName)];
  for (const segment of segments) {
    const found = checkRoleInText(segment, prof);
    if (found) return found;
  }
  return "primary";
}

export function parseQuantity(filename: string, profile?: NamingProfile | null): number {
  const prof = resolveProfile(profile);
  const m = prof.quantityRe.exec(filename);
  if (m?.[1]) return Math.max(1, parseInt(m[1], 10));
  return Math.max(1, prof.quantityDefault);
}

export function parsePartSlug(filename: string, profile?: NamingProfile | null): string {
  const prof = resolveProfile(profile);
  let name = basename(filename);
  if (name.toLowerCase().endsWith(".stl")) name = name.slice(0, -4);
  let stem = name;
  if (prof.slugStripMarkers) {
    for (const prefixRe of prof.rolePrefixRes) {
      stem = stem.replace(prefixRe, "");
    }
  }
  if (prof.slugStripQuantity && prof.quantityStripRe) {
    stem = stem.replace(prof.quantityStripRe, "");
  }
  return stem || name;
}

export function parseStlPath(
  relativePath: string,
  profile?: NamingProfile | null,
): ParsedPart {
  const prof = resolveProfile(profile);
  const filename = basename(relativePath.replace(/\\/g, "/"));
  return {
    role: parseRole(relativePath, prof),
    quantity: parseQuantity(filename, prof),
    partSlug: parsePartSlug(filename, prof),
    filename,
  };
}

export function previewParse(
  relativePath: string,
  profile?: NamingProfile | null,
): { role: string; quantity: number; part_slug: string; filename: string } {
  const parsed = parseStlPath(relativePath, profile ?? undefined);
  return {
    role: parsed.role,
    quantity: parsed.quantity,
    part_slug: parsed.partSlug,
    filename: parsed.filename,
  };
}
